design/gas/Design_gas.md

```diff
@@ -0,0 +1,89 @@
+1) Regular transactions - they work based on a gas auction.
+2) Access list transactions - same as previous ones, but they also announce what storage access they are supposed to perform when executed.
+3) EIP-1559 - the new transactions that burn the gas fee and work with tips in order to achieve a fair system.
+
+Usually layer 2's support EIP-1559 so there is a pending question of what we would need to support. Normally all support EIP-1559, but not all support gas auction (or mention if they do). As most wallets have moved over to EIP-1559 it is preferable to start with support for it and add the support for the gas auction later on after everything else is finished.
+Note that Binance for example still uses Type 1 transactions and has not migrated to EIP-1559. 
+
+## Requirements
+
+1. Ethereum must be used for gas operations and bridged from the layer 1 protocol.
+2. Bridged gas should automatically deposit into the respective addresses.
+3. Gas-consuming transactions should include a layer 1 cost component for calldata publishing and a layer 2 component that covers execution costs and prevents endless transaction execution.
+4. L2 gas price should be configurable. Furthermore it should support being free whilst still not allowing non halting execution.
+5. The gas mechanics need to prevent denial of service attacks.
+6. EIP-1559 transactions should have their base fee + tip deposited to the sequencer without any ETH being burned.
+
+## Gas Cost Display
+
+Metamask can disassemble the cost components for layer 1 and layer 2 for specific transactions on Optimism, such as token balance approvals. However, it remains unclear how they achieve this as it's neither fully documented nor supports all transactions.
+
+
+## Design for Gas Deposit
+
+Given that Ethereum as a currency operates on the layer 1 protocol, it needs to be bridged to Obscuro for usage. This transition mechanism will be facilitated by the standard Obscuro bridge. The deposited value will produce a cross-chain message indicating the Ethereum recipient through the smart contract.
+
+```EthereumDeposited(address receiver, uint256 amount)```
+
+The enclave will automatically pick up this message, along with other cross-chain messages, but it will be treated specially. Upon detection, the enclave will increase the balance of the receiver account by the amount specified in the message. This automatic relaying is needed in order to avoid the chicken and egg problem one would have initially where no relayer has any gas to relay the gas deposit messages. It would also ensure relayers who run out of gas can easily recharge without relying on other relayers if we were to use a different bootstrap mechanism. Note that those deposits will not call the fallback function of the address if it is a smart contract and any deposits to such contracts would later on need a message relayed to verify the message as they would normally.
+
+## Gas Estimation
```

Comment:
This isn't 100% correct, and makes things confusing.

Gas estimation is how much processing power it takes to execute the transaction. 
Which is used for network congestion throttling.
Since :
- The rollup is not executed
- We're not accounting for rollup compression costs
- Once a tx is minted into a batch it's "soft-final"
- We don't have a finality mechanism that allows users to bump their txs to be rolled up faster 

I don't believe there is a point in changing how the gas estimation for a given transaction is done. The binary search tree done in geth should be enough, and I don't see the need to add the L1 calldata cost.

My suggestion would be to leave the gas estimation as geth and move additional payout to the gas price.

Reply:
Gas estimation needs to reflect what the user will actually pay.